[Cocoa] Slack.com audible notifications interrupt Music.app and steal NowPlaying status

jernoble opened the pull request:
#### c320bb08e139b6a698594f0320827767771259fd
<pre>
[Cocoa] Slack.com audible notifications interrupt Music.app and steal NowPlaying status
<a href="https://bugs.webkit.org/show_bug.cgi?id=263750">https://bugs.webkit.org/show_bug.cgi?id=263750</a>
rdar://114667001

Reviewed by Eric Carlson.

Many sites use short-duration audio sources via the &lt;audio&gt; element as
&quot;alert&quot; sounds. When these sounds are played, they are categorized by
existing heuristics as MediaPlayback, and will cause the application to
request NowPlayingApplication status. This breaks playback in other media
playing applications on the system, and prevents hardware playback buttons
from resuming playback from the original application after the interruption.

Add a heuristic for &quot;short duration audio playback&quot; defined by &quot;audio with
a duration of less-than-or-equal-to the AOL youvegotmail.mp3 sound&quot;.

* LayoutTests/media/ios/short-audio-now-playing-expected.txt: Added.
* LayoutTests/media/ios/short-audio-now-playing.html: Added.
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::mediaElementSessionInfoForSession):
(WebCore::HTMLMediaElement::enterFullscreen):
(WebCore::HTMLMediaElement::mediaSessionDuration const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::canShowControlsManager const):
(WebCore::isElementLargeEnoughForMainContent):
(WebCore::isElementLongEnoughForMainContent):
* Source/WebCore/html/MediaElementSession.h:
* Source/WebCore/platform/audio/PlatformMediaSession.cpp:
(WebCore::PlatformMediaSession::isEnded const):
(WebCore::PlatformMediaSession::duration const):
(WebCore::PlatformMediaSessionClient::mediaSessionDuration const):
* Source/WebCore/platform/audio/PlatformMediaSession.h:
(WebCore::PlatformMediaSessionClient::isSuspended const):
(WebCore::PlatformMediaSessionClient::isPlaying const):
(WebCore::PlatformMediaSessionClient::isAudible const):
(WebCore::PlatformMediaSessionClient::isEnded const):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::sessionWillEndPlayback):

Canonical link: <a href="https://commits.webkit.org/269948@main">https://commits.webkit.org/269948@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/84a93b46d2447efd50c15b2527435a7346673d35

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23879 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1991 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26025 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22010 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24153 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3599 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24375 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/22517 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24121 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1536 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20638 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26616 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1291 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21549 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27790 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21769 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21829 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25579 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/ignored-objects (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1244 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18932 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/1252 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/21321 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5768 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1656 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1573 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6793,6 +6795,10 @@ void HTMLMediaElement::enterFullscreen(VideoFullscreenMode mode)
+    auto* page = document.page();
+    if (!page || page->mediaPlaybackIsSuspended())
+        return
+
```
Is this related?

jernoble replied:
lol, no.

eric-carlson commented on Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
```diff
@@ -336,6 +336,24 @@
+    // is played by web content.
+
+    // Derived from the duration of the "You've got mail!" AOL sound:
+    static constexpr MediaTime YouveGotMailDuration = MediaTime(95, 100);
```
If you put `isLongEnoughForMainContent` in `PlatformMediaSession ` you could use it from `MediaElementSession` and not define the constant in more than one place

jernoble committed:
[Cocoa] Slack.com audible notifications interrupt Music.app and steal NowPlaying status
https://bugs.webkit.org/show_bug.cgi?id=263750
rdar://114667001

Reviewed by Eric Carlson.

Many sites use short-duration audio sources via the <audio> element as
"alert" sounds. When these sounds are played, they are categorized by
existing heuristics as MediaPlayback, and will cause the application to
request NowPlayingApplication status. This breaks playback in other media
playing applications on the system, and prevents hardware playback buttons
from resuming playback from the original application after the interruption.

Add a heuristic for "short duration audio playback" defined by "audio with
a duration of less-than-or-equal-to the AOL youvegotmail.mp3 sound".

* LayoutTests/media/ios/short-audio-now-playing-expected.txt: Added.
* LayoutTests/media/ios/short-audio-now-playing.html: Added.
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::mediaElementSessionInfoForSession):
(WebCore::HTMLMediaElement::enterFullscreen):
(WebCore::HTMLMediaElement::mediaSessionDuration const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::canShowControlsManager const):
(WebCore::isElementLargeEnoughForMainContent):
(WebCore::isElementLongEnoughForMainContent):
* Source/WebCore/html/MediaElementSession.h:
* Source/WebCore/platform/audio/PlatformMediaSession.cpp:
(WebCore::PlatformMediaSession::isEnded const):
(WebCore::PlatformMediaSession::duration const):
(WebCore::PlatformMediaSessionClient::mediaSessionDuration const):
* Source/WebCore/platform/audio/PlatformMediaSession.h:
(WebCore::PlatformMediaSessionClient::isSuspended const):
(WebCore::PlatformMediaSessionClient::isPlaying const):
(WebCore::PlatformMediaSessionClient::isAudible const):
(WebCore::PlatformMediaSessionClient::isEnded const):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::sessionWillEndPlayback):

Canonical link: https://commits.webkit.org/269948@main
